packages/h5wasm/src/h5wasm-api.ts

```diff
@@ -0,0 +1,144 @@
+import {
+  assertH5WasmDataset,
+  assertH5WasmEntityWithAttrs,
+  isH5WasmDataset,
+  isH5WasmGroup,
+  isHDF5,
+} from './utils';
+
+export class H5WasmApi extends ProviderApi {
+  private readonly file: Promise<H5WasmFile>;
+
+  public constructor(filename: string, buffer: ArrayBuffer) {
+    super(filename);
+
+    if (!isHDF5(buffer)) {
+      throw new Error('Expected valid HDF5 file');
+    }
+
+    this.file = this.initFile(buffer);
+  }
+
+  public async getEntity(path: string): Promise<Entity> {
+    const file = await this.file;
+    const h5wEntity = file.get(path);
+
+    const segments = path.slice(1).split('/');
+    return this.processH5WasmEntity(
+      segments[segments.length - 1],
+      path,
+      h5wEntity
+    );
+  }
+
+  public async getValue(params: ValuesStoreParams): Promise<unknown> {
+    const { dataset } = params;
+
+    const file = await this.file;
+    const h5wDataset = file.get(dataset.path);
+    assertH5WasmDataset(h5wDataset);
+
+    const array = h5wDataset.value;
+    return hasScalarShape(dataset) ? array[0] : array;
+  }
+
+  public async getAttrValues(entity: Entity): Promise<AttributeValues> {
+    const file = await this.file;
+    const h5wEntity = file.get(entity.path);
+    assertH5WasmEntityWithAttrs(h5wEntity);
+
+    return Object.fromEntries(
+      Object.entries(h5wEntity.attrs).map(([name, attr]) => {
+        return [name, (attr as H5WasmAttribute).value];
+      })
+    );
+  }
+
+  public async cleanUp(): Promise<void> {
+    const file = await this.file;
+    file.close();
+  }
+
+  private async initFile(buffer: ArrayBuffer): Promise<H5WasmFile> {
+    await ready;
+
+    FS.writeFile(this.filepath, new Uint8Array(buffer), { flags: 'w+' });
+    return new H5WasmFile(this.filepath, 'r');
+  }
+
+  private processH5WasmEntity(
+    name: string,
+    path: string,
+    h5wEntity: H5WasmEntity,
+    isChild = false
+  ): Entity {
+    const baseEntity = { name, path };
+
+    if (isH5WasmGroup(h5wEntity)) {
+      const baseGroup: Group = {
+        ...baseEntity,
+        kind: EntityKind.Group,
+        attributes: this.processH5WasmAttrs(h5wEntity.attrs),
+      };
+
+      if (isChild) {
+        return baseGroup;
+      }
+
+      const children = h5wEntity.keys().map((name) => {
+        const h5wChild = h5wEntity.get(name);
+        const childPath = buildEntityPath(path, name);
+        return this.processH5WasmEntity(name, childPath, h5wChild, true);
+      });
+
+      return { ...baseGroup, children } as GroupWithChildren;
+    }
+
+    if (isH5WasmDataset(h5wEntity)) {
+      const { shape, dtype } = h5wEntity;
+
+      return {
+        ...baseEntity,
+        kind: EntityKind.Dataset,
+        attributes: this.processH5WasmAttrs(h5wEntity.attrs),
+        shape,
+        type:
+          typeof dtype === 'string'
+            ? convertDtype(dtype)
+            : { class: DTypeClass.Compound, fields: dtype.compound },
```

Comment:
h5wasm dtypes masy need specific conversions so we will have to revisit this later.

Reply:
These conversions were implemented in https://github.com/silx-kit/h5web/pull/1031/commits/f202aa8c604afd8cd2a029613e48429ea790f661 if you want to port them in another PR.